runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -252,130 +221,100 @@ Ref<PredictionContext> PredictionContext::mergeArrays(const Ref<ArrayPredictionC
   const Ref<ArrayPredictionContext> &b, bool rootIsWildcard, PredictionContextMergeCache *mergeCache) {
 
   if (mergeCache != nullptr) {
-    auto existing = mergeCache->get(a, b);
-    if (existing) {
-      return existing;
-    }
-    existing = mergeCache->get(b, a);
-    if (existing) {
+    if (const Ref<PredictionContext>& existing = mergeCache->get(a, b)) {
       return existing;
     }
   }
 
   // merge sorted payloads a + b => M
-  size_t i = 0; // walks a
-  size_t j = 0; // walks b
-  size_t k = 0; // walks target M array
-
-  std::vector<size_t> mergedReturnStates(a->returnStates.size() + b->returnStates.size());
-  std::vector<Ref<PredictionContext>> mergedParents(a->returnStates.size() + b->returnStates.size());
-
-  // walk and merge to yield mergedParents, mergedReturnStates
-  while (i < a->returnStates.size() && j < b->returnStates.size()) {
-    Ref<PredictionContext> a_parent = a->parents[i];
-    Ref<PredictionContext> b_parent = b->parents[j];
-    if (a->returnStates[i] == b->returnStates[j]) {
-      // same payload (stack tops are equal), must yield merged singleton
-      size_t payload = a->returnStates[i];
-      // $+$ = $
-      bool both$ = payload == EMPTY_RETURN_STATE && a_parent && b_parent;
-      bool ax_ax = (a_parent && b_parent) && *a_parent == *b_parent; // ax+ax -> ax
-      if (both$ || ax_ax) {
-        mergedParents[k] = a_parent; // choose left
-        mergedReturnStates[k] = payload;
-      }
-      else { // ax+ay -> a'[x,y]
-        Ref<PredictionContext> mergedParent = merge(a_parent, b_parent, rootIsWildcard, mergeCache);
-        mergedParents[k] = mergedParent;
-        mergedReturnStates[k] = payload;
+  std::vector<PredictionContextItem> contexts;
+  contexts.reserve(a->contexts.size() + b->contexts.size());
+  std::vector<PredictionContextItem>::const_iterator i = a->contexts.begin(), i_end = a->contexts.end();
+  std::vector<PredictionContextItem>::const_iterator j = b->contexts.begin(), j_end = b->contexts.end();
+  while (i != i_end && j != j_end) {
```

Comment:
Looks good, however, can you save the original comments, at least those that describe the operation (e.g. `// ax+ay -> a'[x,y]`)?

Reply:
Have put these comments back in and added the same to mergeSingletonIntoArray as well. I also modified mergeSingletonIntoArray so it should be doing exactly the same for a single item context as mergeArrays does not a many item context. In particular for empty the parents of a merged item don't need to be the same.